java/code/src/com/redhat/rhn/common/util/FileUtils.java

```diff
@@ -61,12 +61,13 @@ public static void writeStringToFile(String contents, String path) {
                 file.delete();
             }
             file.createNewFile();
-            Writer output = new BufferedWriter(new FileWriter(file));
-            try {
-              output.write(contents);
-            }
-            finally {
-              output.close();
+            try (FileOutputStream fos = new FileOutputStream(file);
+                 FileWriter fw = new FileWriter(fos.getFD());
+                 BufferedWriter output = new BufferedWriter(fw)) {
+                output.write(contents);
+                output.flush();
+                fw.flush();
+                fos.getFD().sync();
```

Comment:
Wow, I wouldn't have imagined `FileWriter.flush()` did **not** equate to an underlying `sync` at the OS level!

It'd be good to disseminate this tiny bit of knowledge (via an email or Review), as I can imagine many other Java developers not knowing.

I also wonder whether we might have other notable cases. Pillars maybe, @meaksh?

Reply:
Yeah, I've seen some other places at our Java codebase that are using `FileWriter` in a way it's not explicitly syncing with the OS.

I don't know if this is causing issues already but it might potentially cause them in some particular scenarios.

```l
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:import java.io.FileWriter;
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:        try (BufferedWriter writer = new BufferedWriter(new FileWriter(dataFile))) {
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:        try (BufferedWriter writer = new BufferedWriter(new FileWriter(dataFile))) {
code/src/com/redhat/rhn/domain/formula/FormulaFactory.java:        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
code/src/com/redhat/rhn/manager/audit/AuditManager.java:import java.io.FileWriter;
code/src/com/redhat/rhn/manager/audit/AuditManager.java:        FileWriter fwr = new FileWriter(reviewFile, true); // append!
code/src/com/redhat/rhn/taskomatic/task/repomd/DebPackageWriter.java:import java.io.FileWriter;
code/src/com/redhat/rhn/taskomatic/task/repomd/DebPackageWriter.java:        out = new BufferedWriter(new FileWriter(
code/src/com/redhat/rhn/taskomatic/task/repomd/RpmRepositoryWriter.java:import java.io.FileWriter;
code/src/com/redhat/rhn/taskomatic/task/repomd/RpmRepositoryWriter.java:        FileWriter indexFile;
code/src/com/redhat/rhn/taskomatic/task/repomd/RpmRepositoryWriter.java:            indexFile = new FileWriter(prefix + REPOMD_FILE);
code/src/com/redhat/rhn/taskomatic/task/repomd/RpmRepositoryWriter.java:            FileWriter norepo = new FileWriter(prefix + NOREPO_FILE);
code/src/com/redhat/rhn/taskomatic/domain/TaskoRun.java:import java.io.FileWriter;
code/src/com/redhat/rhn/taskomatic/domain/TaskoRun.java:            BufferedWriter out = new BufferedWriter(new FileWriter(fileName, true));
code/src/com/suse/manager/webui/services/SaltActionChainGeneratorService.java:import java.io.FileWriter;
code/src/com/suse/manager/webui/services/SaltActionChainGeneratorService.java:        try (Writer slsWriter = new FileWriter(targetFilePath.toFile());
code/src/com/suse/manager/webui/services/SaltActionChainGeneratorService.java:        try (Writer fout = new FileWriter(topPath.toFile())) {
code/src/com/suse/manager/webui/utils/SaltStateGenerator.java:import java.io.FileWriter;
code/src/com/suse/manager/webui/utils/SaltStateGenerator.java:        this.destination = new FileWriter(destinationIn);
```

And yeah, it seems pillar generator, via SaltStateGenerator, might be affected as well. Also formulas, actions chains, etc.

I think we should create a card to review and fix those usages of the `FileWriter` in the rest of the codebase in order to be really on the safe side. I'll create it.

About spreading the knowledge, I'll send an email to our devel mailing list and maybe also present something about this on next review meeting.